File: src/part2/Lim10844.java
package part2;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Lim10844 {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		
		long[] cnt = {0, 1, 1, 1, 1, 1, 1, 1, 1, 1};
		
		for(int i = 1; i < n; i++) {
			long[] tmp = new long[10];
			
			tmp[0] = cnt[1];
			
			for(int j = 1; j <= 8; j++) {
				tmp[j] = (cnt[j - 1] + cnt[ j + 1]) % 1_000_000_000;
			}
			
			tmp[9] = cnt[8];
			
			cnt = tmp;
		}
		
		long sum = 0;
		
		for(int i = 0; i < 10; i++) {
			sum += cnt[i];
		}
		
		System.out.println(sum % 1_000_000_000);
	}
}

File: src/main/java/boj/workbook/dp/Lim10844.kt
package boj.workbook.dp

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    val dp = Array(10) { IntArray(101) }
    for (i in 1..9) {
        dp[i][1] = 1
    }

    for (i in 2..n) {
        dp[0][i] = dp[1][i - 1]
        for (j in 1..8) {
            dp[j][i] = (dp[j - 1][i - 1] + dp[j + 1][i - 1]) % 1_000_000_000
        }
        dp[9][i] = dp[8][i - 1]
    }

    var answer = 0
    for (i in 0..9) {
        answer = (answer + dp[i][n]) % 1_000_000_000
    }
    println(answer)
}
